src/GpuAsMailboxprocessorDemo/EWiseAdd.fs

```diff
@@ -0,0 +1,122 @@
+namespace GraphBLAS.FSharp.Backend.COOMatrix
+
+open Brahma.FSharp.OpenCL
+open OpenCL.Net
+open GraphBLAS.FSharp.Backend.Common
+open GraphBLAS.FSharp.Backend.COOMatrix.Utilities
+open Microsoft.FSharp.Quotations
+
+module internal rec EWiseAdd =
+    let run (gpu:GPU) (matrixLeft: COOMatrix<'a>) (matrixRight: COOMatrix<'a>) (op: Expr<'a -> 'a -> 'a>) =
+        if matrixLeft.Values.Length = 0 then
+            let resultRows = matrixRight.Rows
+            let resultColumns = matrixRight.Columns
+            let resultValues = matrixRight.Values
+
+            (0L,0L,0L)
+            , {
+                RowCount = matrixRight.RowCount
+                ColumnCount = matrixRight.ColumnCount
+                Rows = resultRows
+                Columns = resultColumns
+                Values = resultValues
+              }
+
+        elif matrixRight.Values.Length = 0 then
+            let resultRows = matrixLeft.Rows
+            let resultColumns = matrixLeft.Columns
+            let resultValues = matrixLeft.Values
+
+            (0L,0L,0L)
+            , {
+                RowCount = matrixLeft.RowCount
+                ColumnCount = matrixLeft.ColumnCount
+                Rows = resultRows
+                Columns = resultColumns
+                Values = resultValues
+              }
+
+        else
+            let _do = gpuEWiseAdd gpu op
+            _do matrixLeft matrixRight 
```

Comment:
Am I right that every time ```run``` is called, the kernels will compile again? So, ```run``` must be a function which call ```gpuEWiseAdd``` and then return a function, isn't it?

Reply:
You are right. ```EWiseAdd``` function should be redesigned. I guess, in order to avoid this problem and in order to be composable, it should be designed similar to other functions in the example (```prefixSum```, ```setPositions```, etc)